File: packages/verifier-ts/src/canonical.ts
// RFC 8785 JCS canonicalizer for Uniclaw receipt bodies.
//
// This is the TypeScript port of `crates/uniclaw-receipt/src/canonical.rs`.
// The browser verifier (`crates/uniclaw-host/src/verify.html`) and
// the Node conformance smoke
// (`crates/uniclaw-receipt/tests/vectors/conformance-smoke.mjs`)
// embed equivalent JS implementations; this package is the
// canonical TypeScript reference.
//
// Cross-language byte-identity is enforced by
// `tests/conformance.test.ts`, which loads the same fixture file
// (`canonical-v2.json`) the Rust snapshot test loads. If the
// fixture passes here AND in Rust, both implementations agree.
//
// Scope: handles the shape Uniclaw receipts actually use — string
// keys, integer numbers only, ASCII-mostly strings. Floats throw
// rather than emit potentially wrong bytes (matches Rust behavior,
// since the schema has no float fields and a future addition must
// update both canonicalizers in lockstep).

import type { JsonValue, ReceiptBody } from "./types.js";

// Encode `body` to canonical bytes the kernel signed.
//
// Dispatches on `body.schema_version`:
//   - schema_version <= 1: use the legacy `JSON.stringify(body)`
//     path. ES2015 preserves insertion order, and a body re-parsed
//     from JSON-as-served-by-the-host keeps Rust's struct-declaration
//     order — so the bytes match.
//   - schema_version >= 2: use RFC 8785 JCS (lexicographic key
//     sort, normalized integers, standard string escapes).
export function canonicalizeBody(body: ReceiptBody): Uint8Array {
  const str =
    typeof body === "object" && body !== null && (body.schema_version ?? 0) >= 2
      ? canonicalizeJcs(body as unknown as JsonValue)
      : JSON.stringify(body);
  return new TextEncoder().encode(str);
}

// Canonicalize an arbitrary JSON value using RFC 8785 JCS rules.
// Exported for tests and for callers who want to canonicalize a
// fragment for debugging.
export function canonicalizeJcs(value: JsonValue): string {
  if (value === null) return "null";
  if (value === true) return "true";
  if (value === false) return "false";
  if (typeof value === "number") {
    if (!Number.isInteger(value)) {
      throw new Error(`JCS: expected integer, got float ${value}`);
    }
    return String(value);
  }
  if (typeof value === "string") {
    return canonicalizeJcsString(value);
  }
  if (Array.isArray(value)) {
    return "[" + value.map(canonicalizeJcs).join(",") + "]";
  }
  if (typeof value === "object") {
    // Sort keys by UTF-16 code unit order. JS's default `sort()` on
    // strings does abstract relational comparison, which is the
    // UTF-16 code unit ordering JCS specifies.
    const keys = Object.keys(value).sort();
    return (
      "{" +
      keys
        .map((k) => {
          const v = (value as { [k: string]: JsonValue })[k];
          // `noUncheckedIndexedAccess` widens to `JsonValue | undefined`;
          // we just listed `k` from `Object.keys`, so it's defined.
          if (v === undefined) {
            throw new Error(`JCS: key disappeared during sort: ${k}`);
          }
          return canonicalizeJcsString(k) + ":" + canonicalizeJcs(v);
        })
        .join(",") +
      "}"
    );
  }
  throw new Error(`JCS: unsupported value type ${typeof value}`);
}

function canonicalizeJcsString(s: string): string {
  let out = '"';
  for (const c of s) {
    switch (c) {
      case '"':
        out += '\\"';
        break;
      case "\\":
        out += "\\\\";
        break;
      case "\b":
        out += "\\b";
        break;
      case "\f":
        out += "\\f";
        break;
      case "\n":
        out += "\\n";
        break;
      case "\r":
        out += "\\r";
        break;
      case "\t":
        out += "\\t";
        break;
      default: {
        const code = c.codePointAt(0);
        if (code === undefined) {
          throw new Error("JCS: empty code point");
        }
        if (code < 0x20) {
          out += "\\u" + code.toString(16).padStart(4, "0");
        } else {
          out += c;
        }
      }
    }
  }
  return out + '"';
}

File: packages/client-py/src/uniclaw_client/_canonical.py
"""RFC 8785 JCS canonicalizer for Uniclaw receipt bodies.

Python port of ``crates/uniclaw-receipt/src/canonical.rs`` and
``packages/verifier-ts/src/canonical.ts``. The three implementations
share a single conformance fixture
(``crates/uniclaw-receipt/tests/vectors/canonical-v2.json``); any
divergence fails the conformance test in whichever language drifted.

Scope: handles the shape Uniclaw receipts use — string keys,
integer numbers only, ASCII-mostly strings. Floats raise rather than
emit potentially wrong bytes (matches Rust and TS behavior).
"""

from __future__ import annotations

from typing import Any


def canonicalize_body(body: dict[str, Any]) -> bytes:
    """Encode a receipt body to canonical bytes.

    Dispatches on ``body.schema_version``:

    - ``<= 1``: legacy ``json.dumps`` (struct-declaration order is
      preserved when the body was already parsed from JSON text).
    - ``>= 2``: RFC 8785 JCS (lexicographic key sort, integer
      numbers only, standard string escapes).
    """
    if not isinstance(body, dict):
        raise TypeError("body must be a dict")
    schema_version = body.get("schema_version", 0)
    if isinstance(schema_version, int) and schema_version >= 2:
        return canonicalize_jcs(body).encode("utf-8")
    # Legacy path: json.dumps with insertion-order keys + no spaces.
    # Importing here keeps the JCS path free of stdlib JSON overhead.
    import json

    return json.dumps(body, separators=(",", ":"), ensure_ascii=False).encode("utf-8")


def canonicalize_jcs(value: Any) -> str:
    """Canonicalize an arbitrary JSON value to a string per RFC 8785."""
    if value is None:
        return "null"
    if value is True:
        return "true"
    if value is False:
        return "false"
    if isinstance(value, bool):
        # Unreachable — bool is True/False above. Defensive belt.
        raise RuntimeError("bool fell through")
    if isinstance(value, int):
        return str(value)
    if isinstance(value, float):
        raise ValueError(f"JCS: expected integer, got float {value!r}")
    if isinstance(value, str):
        return _canonicalize_jcs_string(value)
    if isinstance(value, list):
        return "[" + ",".join(canonicalize_jcs(v) for v in value) + "]"
    if isinstance(value, dict):
        # Sort keys by UTF-16 code-unit order. Python's str default
        # ordering is by codepoint; for the BMP-only strings Uniclaw
        # uses today, that matches UTF-16 ordering byte-for-byte.
        # If a future field admits non-BMP keys, this function will
        # need a real UTF-16 comparison; the conformance fixture
        # would catch that drift the same way Rust + TS do.
        keys = sorted(value.keys())
        parts: list[str] = []
        for k in keys:
            if not isinstance(k, str):
                raise TypeError(f"JCS: object key must be str, got {type(k).__name__}")
            parts.append(_canonicalize_jcs_string(k) + ":" + canonicalize_jcs(value[k]))
        return "{" + ",".join(parts) + "}"
    raise TypeError(f"JCS: unsupported value type {type(value).__name__}")


def _canonicalize_jcs_string(s: str) -> str:
    out = ['"']
    for ch in s:
        code = ord(ch)
        if ch == '"':
            out.append('\\"')
        elif ch == "\\":
            out.append("\\\\")
        elif ch == "\b":
            out.append("\\b")
        elif ch == "\f":
            out.append("\\f")
        elif ch == "\n":
            out.append("\\n")
        elif ch == "\r":
            out.append("\\r")
        elif ch == "\t":
            out.append("\\t")
        elif code < 0x20:
            out.append(f"\\u{code:04x}")
        else:
            out.append(ch)
    out.append('"')
    return "".join(out)
